app/models/user.rb

```diff
@@ -492,6 +503,16 @@ def guard_unconfirmed_email
     end
   end
 
+  # Whenever email is set, clear all unconfirmed emails which match
+  def remove_invalid_unconfirmed_emails
+    if email_changed?
+      User.where("unconfirmed_email = ?", email).find_each do |problem_user|
+        problem_user.unconfirmed_email = nil
+        problem_user.save
+      end
+    end
+  end
```

Comment:
I don't think setting it to `nil` can ever make a user fail validation, so `User.where(unconfirmed_email: email).update_all(unconfirmed_email: nil)` should be much, much faster.